File: src/stages/stage4.ts
import path from "node:path";

import { runAgent } from "../agents/index.js";
import { CheckpointManager } from "../checkpoint.js";
import type { AuditConfig, EntryPoint } from "../config.js";
import { getLogger } from "../logger.js";
import { loadPrompt } from "../prompts.js";
import { coerceError, formatValidationIssues, listMatchingFiles, runParallelLimited } from "../utils.js";
import { validateStage4File } from "../validation/stage4.js";

const logger = getLogger("stage4");

function taskKey(entryPoint: EntryPoint): string {
  return `stage4:${entryPoint.moduleId}:${entryPoint.id}`;
}

async function runEntryPoint(
  entryPoint: EntryPoint,
  config: AuditConfig,
  checkpoint: CheckpointManager,
): Promise<string[]> {
  const key = taskKey(entryPoint);
  const resultDir = path.join(config.outputDir, "stage-4-details");
  const findingPrefix = `${entryPoint.moduleId}-${entryPoint.id}`;
  const findingPattern = new RegExp(`^${findingPrefix}-F-\\d+\\.md$`);

  if (checkpoint.isComplete(key)) {
    logger.info("Stage 4: %s/%s already complete, skipping.", entryPoint.moduleId, entryPoint.id);
    return await listMatchingFiles(resultDir, findingPattern);
  }

  const prompt = await loadPrompt("stage4.md", {
    ep_block: entryPoint.rawBlock,
    module_id: entryPoint.moduleId,
    ep_id: entryPoint.id,
    ep_type: entryPoint.type,
    location: entryPoint.location,
    attacker_controlled_data: entryPoint.attackerControlledData,
    initial_validation: entryPoint.initialValidation || "None observed",
    analysis_hints: entryPoint.analysisHints,
    result_dir: resultDir,
    finding_prefix: findingPrefix,
    target_path: config.target,
  });

  await runAgent({
    prompt,
    config,
    cwd: config.target,
  });

  const findingFiles = await listMatchingFiles(resultDir, findingPattern);
  for (const findingFile of findingFiles) {
    let issues = validateStage4File(findingFile);
    if (issues.length === 0) {
      continue;
    }

    logger.warning("Stage 4: Validation failed for %s\n%s", findingFile, formatValidationIssues(issues));
    const repairPrompt =
      `The finding file at \`${findingFile}\` failed validation. ` +
      `Please fix all issues listed below:\n\n\`\`\`\n${formatValidationIssues(issues)}\n\`\`\``;

    await runAgent({
      prompt: repairPrompt,
      config,
      cwd: config.target,
      maxTurns: 10,
    });

    issues = validateStage4File(findingFile);
    if (issues.length > 0) {
      logger.warning("Stage 4: Repair failed for %s\n%s", findingFile, formatValidationIssues(issues));
    }
  }

  checkpoint.markComplete(key);
  logger.info("Stage 4: %s/%s complete. Findings: %s", entryPoint.moduleId, entryPoint.id, findingFiles.length);
  return findingFiles;
}

export async function runStage4(
  entryPointMap: Record<string, EntryPoint[]>,
  config: AuditConfig,
  checkpoint: CheckpointManager,
): Promise<string[]> {
  const allEntryPoints = Object.values(entryPointMap).flat();
  if (allEntryPoints.length === 0) {
    logger.warning("Stage 4: No entry points to analyze.");
    return [];
  }

  const results = await runParallelLimited(allEntryPoints, config.maxParallel, async (entryPoint) => {
    return await runEntryPoint(entryPoint, config, checkpoint);
  });

  const allFindingFiles: string[] = [];
  results.forEach((result, index) => {
    const entryPoint = allEntryPoints[index];
    if (!entryPoint) {
      return;
    }
    if (result.status === "rejected") {
      logger.error(
        "Stage 4: %s/%s failed with exception: %s",
        entryPoint.moduleId,
        entryPoint.id,
        coerceError(result.reason).message,
      );
      return;
    }
    allFindingFiles.push(...result.value);
  });

  logger.info("Stage 4 complete. Total finding files: %s", allFindingFiles.length);
  return allFindingFiles;
}

File: code_auditor/stages/stage4.py
from __future__ import annotations

import os
import re

from ..agent import run_agent
from ..checkpoint import CheckpointManager
from ..config import AnalysisUnit, AuditConfig
from ..logger import get_logger
from ..prompts import load_prompt
from ..utils import format_validation_issues, list_matching_files, run_parallel_limited
from ..validation.stage4 import validate_stage4_file

logger = get_logger("stage4")


def _task_key(unit: AnalysisUnit) -> str:
    return f"stage4:{unit.id}"


async def _run_unit(
    unit: AnalysisUnit,
    config: AuditConfig,
    checkpoint: CheckpointManager,
    auditing_focus_path: str,
    vuln_criteria_path: str,
    deployment_summary_path: str,
    unit_index: int = 0,
    total_units: int = 0,
) -> list[str]:
    key = _task_key(unit)
    result_dir = os.path.join(config.output_dir, "stage4-findings")
    log_file = os.path.join(result_dir, "logs", f"{unit.id}.log")
    escaped_id = re.escape(unit.id)
    finding_pattern = re.compile(rf"^{escaped_id}-F-\d+\.json$")
    progress = f"[{unit_index}/{total_units}]" if total_units else ""

    if checkpoint.is_complete(key):
        logger.info("Stage 4 %s: %s already complete, skipping.", progress, unit.id)
        return list_matching_files(result_dir, finding_pattern)

    logger.info("Stage 4 %s: Starting bug discovery for %s.", progress, unit.id)
    deployment_summary = ""
    if os.path.exists(deployment_summary_path):
        with open(deployment_summary_path) as f:
            deployment_summary = f.read().strip()

    prompt = load_prompt("stage4.md", {
        "au_file_path": unit.au_file_path,
        "result_dir": result_dir,
        "finding_prefix": unit.id,
        "auditing_focus_path": auditing_focus_path,
        "vuln_criteria_path": vuln_criteria_path,
        "deployment_summary": deployment_summary or "No deployment summary available.",
    })

    await run_agent(prompt, config, cwd=config.target, max_turns=200, log_file=log_file)

    logger.info("Stage 4 %s: Agent finished for %s. Validating findings.", progress, unit.id)
    finding_files = list_matching_files(result_dir, finding_pattern)
    for finding_file in finding_files:
        issues = validate_stage4_file(finding_file)
        if not issues:
            continue

        logger.warning("Stage 4: Validation failed for %s\n%s", finding_file, format_validation_issues(issues))
        repair_prompt = (
            f"The finding file at `{finding_file}` failed validation. "
            f"Please fix all issues listed below:\n\n```\n{format_validation_issues(issues)}\n```"
        )
        await run_agent(repair_prompt, config, cwd=config.target, max_turns=10, log_file=log_file)

        issues = validate_stage4_file(finding_file)
        if issues:
            logger.warning("Stage 4: Repair failed for %s\n%s", finding_file, format_validation_issues(issues))

    checkpoint.mark_complete(key)
    logger.info("Stage 4 %s: %s complete. Findings: %d", progress, unit.id, len(finding_files))
    return finding_files


async def run_stage4(
    units: list[AnalysisUnit],
    config: AuditConfig,
    checkpoint: CheckpointManager,
    auditing_focus_path: str,
    vuln_criteria_path: str,
    deployment_summary_path: str,
) -> list[str]:
    if not units:
        logger.warning("Stage 4: No analysis units to process.")
        return []

    total = len(units)
    logger.info("Stage 4: Starting bug discovery across %d analysis units (max parallel: %d).", total, config.max_parallel)

    results = await run_parallel_limited(
        units,
        config.max_parallel,
        lambda unit, idx: _run_unit(
            unit, config, checkpoint, auditing_focus_path, vuln_criteria_path,
            deployment_summary_path,
            unit_index=idx + 1, total_units=total,
        ),
    )

    all_finding_files: list[str] = []
    for i, (status, value, error) in enumerate(results):
        if i >= len(units):
            continue
        if status == "rejected":
            logger.error("Stage 4: %s failed: %s", units[i].id, error)
            continue
        if value:
            all_finding_files.extend(value)

    logger.info("Stage 4 complete. Total bug finding files: %s", len(all_finding_files))
    return all_finding_files
